mbuild/conversion.py

```diff
@@ -0,0 +1,1364 @@
+             'residue templates')
+        structure = pmd.load_file(filename, structure=True, **kwargs)
+        compound = from_parmed(
+                            structure=structure,
+                            compound=compound,
+                            coords_only=coords_only,
+                            infer_hierarchy=infer_hierarchy)
+
+    # Then mdtraj reader
+    elif backend == 'mdtraj':
+        traj = md.load(filename, **kwargs)
+        compound = from_trajectory(
+                            traj=traj,
+                            compound=compound,
+                            frame=-1,
+                            coords_only=coords_only,
+                            infer_hierarchy=infer_hierarchy)
+
+    if rigid:
+        compound.label_rigid_bodies()
+
+    # Note: 'Input not supported' error will be handled
+    # by the corresponding backend
+    return compound
+
+
+def from_parmed(structure, compound=None, coords_only=False,
+                infer_hierarchy=True):
+    """ Backend-specific loading function - parmed
+
+    Parameters
+    ----------
+    structure : pmd.Structure
+        New structure to be loaded.
+    compound : mb.Compound, optional, default=None
+        Host mb.Compound that we are loading to.
+    coords_only : bool, optional, default=False
+        Set preexisting atoms in compound to coordinates
+        given by structure.
+    infer_hierarchy : bool , optional, default=True
+        If True, infer compound hierarchy from chain
+        and reisdues
+
+    Returns
+    -------
+    compound : mb.Compound
+
+    """
+    # Check coords_only option
+    if compound and coords_only:
+        if len(structure.atoms) != compound.n_particles:
+            raise ValueError(
+                'Number of atoms in {structure} does not '
+                '{compound}'.formats(**locals()))
+        atoms_particles = zip(structure.atoms,
+                    compound.particles(include_ports=False))
+        if None in compound._particles(include_ports=False):
+            raise ValueError('Some particles are None')
+        for pmd_atom, particle in atoms_particles:
+            particle.pos = np.array([pmd_atom.xx,
+                                     pmd_atom.xy,
+                                     pmd_atom.xz]) / 10
+        return compound
+    elif not compound and coords_only:
+        raise MBuildError('coords_only=True but'
+                'host compound is not provided')
+
+    # Initialize a compound if none is provided
+    if not compound:
+        compound = mb.Compound()
+
+    # Convert parmed structure to mbuild compound
+    atom_mapping = dict()
+    chain_id = None
+    chains = defaultdict(list)
+
+    # Build up chains dict
+    # ?Could I change this to normal dict?
+    for residue in structure.residues:
+        chains[residue.chain].append(residue)
+
+    # Build up compound
+    for chain, residues in chains.items():
```

Comment:
Have we tested this capability on a complex molecule that has residues and chains? Maybe a large protein PDB file?